### PR TITLE
Create the draw.io GH Action

### PR DESCRIPTION
**Pull request by luxas**
Let's see if it works!

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+  build_image:
+    runs-on: ubuntu-latest
+    name: "Build and push image"
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Log in to ghcr.io
+      run: "echo ${{ secrets.RACKLET_CONTAINERS_GHCR }} | docker login ghcr.io -u $GITHUB_ACTOR --password-stdin"
+    - name: Download tags, for use in the Makefile
```
Comment (luxas) [inline .github/workflows/build.yml:12]: Instead of doing this; I should be able to just do a unshallow tag pull manually

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.16-alpine as builder
+WORKDIR /build
+COPY main.go go.mod go.sum ./
+RUN CGO_ENABLED=0 go build -a -o drawio-renderer .
+
+
+FROM node:lts-alpine3.13
+# Enables README support, etc. in Github Packages. See: https://docs.github.com/en/packages/guides/about-github-container-registry
+LABEL org.opencontainers.image.source https://github.com/racklet/drawio-render-action
+RUN apk add chromium
```
Comment (luxas) [inline Dockerfile:10]: Add attibution to the original project used here maybe?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+on: [push]
```
Comment (luxas) [inline .github/workflows/build.yml:1]: change this to only run tags/releases

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+build:
+	docker build -t $(IMAGE):$(GIT_TAG) --pull .
+
+push: build
+	docker push $(IMAGE):$(GIT_TAG)
+ifneq ($(TAG),)
+	docker tag $(IMAGE):$(GIT_TAG) $(IMAGE):$(TAG)
+	docker push $(IMAGE):$(TAG)
+endif
+ifeq ($(GIT_STATUS),)
```
Comment (luxas) [inline Makefile:26]: split this out to a dedicated `push-major` tag, to not override the "stable" tag for e.g. main branch pushes?

File: README.md [hunk 5]
```diff
@@ -1,29 +1,123 @@
...
 
 Other interesting resources include:
 
-- [The issue tracker](https://github.com/racklet/racklet/issues)
-- [The discussions forum](https://github.com/racklet/racklet/discussions)
-- [The list of milestones](https://github.com/racklet/racklet/milestones)
-- _The roadmap (TODO)_
-- _The changelog (TODO)_
+- [The issue tracker](https://github.com/racklet/drawio-render-action/issues)
+- [The changelog](CHANGELOG.md)
```
Comment (luxas) [inline README.md:112]: actually add a changelog

File: main.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	// Setup default config
+	cfg := &Config{
+		RootDir: "/files",
+		Formats: []string{"png", "pdf"},
+		SubDirs: []string{"."},
+		Files:   map[string]string{},
+	}
+
+	// Unconditionally overwrite root path if GH Action workspace is set
+	if ghWorkspace := os.Getenv("GITHUB_WORKSPACE"); len(ghWorkspace) != 0 {
```
Comment (luxas) [inline main.go:45]: maybe we can just pass this param from the action, to avoid having this "hack" in the code?

File: main.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+// render expects a valid config
+func render(cfg *Config) error {
+	log := zap.S()
+	log.Infow("Got config", "cfg", cfg)
+
+	osfs := os.DirFS(cfg.RootDir)
+	log.Infof("Created os.DirFS at %s", cfg.RootDir)
+
+	for _, subDir := range cfg.SubDirs {
```
Comment (luxas) [inline main.go:150]: guess this should be called `Config.Complete`

File: main.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to run drawio for src=%q and dest=%q: %v, output: %s", absSrc, absDest, err, string(out))
+		}
+		// The output file does not include the root directory prefix
+		outputFiles = append(outputFiles, dest)
+	}
+
+	log.Infow("Got output files", "output-files", outputFiles)
+	// Deliberately include the newline
+	fmt.Printf(`::set-output name=RENDERED_FILES::%s
```
Comment (luxas) [inline main.go:199]: todo: rename this to `rendered-files`

Comment (luxas) [reply main.go:199]: and make an util func

File: main.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	"strings"
+
+	"github.com/scylladb/go-set/strset"
+	"github.com/spf13/pflag"
+	"go.uber.org/zap"
+)
+
+var (
+	validFormats = strset.New("png", "pdf")
+	skipDirs     = strset.New(".git")
```
Comment (luxas) [inline main.go:19]: make configurable?

File: main.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func addFlags(cfg *Config) {
+	pflag.StringVarP(&cfg.RootDir, "root-dir", "r", cfg.RootDir, "Where the root directory for the files that should be converted are. If the env var GITHUB_WORKSPACE is set, this flag will automatically be set to that value.")
+	pflag.StringSliceVar(&cfg.Formats, "formats", cfg.Formats, "Comma-separated list of formats to render the *.drawio files as, for use with --subdirs")
+	pflag.StringSliceVarP(&cfg.SubDirs, "subdirs", "d", cfg.SubDirs, "Comma-separated list of sub-directories of --root-dir to recursively search for *.drawio files to convert")
+	pflag.StringToStringVarP(&cfg.Files, "files", "f", cfg.Files, fmt.Sprintf("Comma-separated list of files to convert, of form 'dest-file=src-file'. The extension for src-file can be anything, while the dest-file extension must be either %s", validFormats))
+}
+
+type Config struct {
```
Comment (luxas) [inline main.go:69]: extract into a separate package, so that it can be re-used by e.g. a plantuml generator?

File: README.md [hunk 11]
```diff
@@ -1,29 +1,123 @@
-# base-repo-layout
+# drawio-render-action
```
Comment (luxas) [inline README.md:1]: later add "official" GH action badge?

File: .github/workflows/e2e.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    runs-on: ubuntu-latest
+    name: Test rendering draw.io files
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Render draw.io files
+      uses: ./
+      id: render
+    - name: Print the rendered files
+      run: echo "The files that changed were ${{ steps.render.outputs.RENDERED_FILES }}"
```
Comment (luxas) [inline .github/workflows/e2e.yml:14]: rename to rendered-files

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.16-alpine as builder
+WORKDIR /build
+COPY main.go go.mod go.sum ./
+RUN CGO_ENABLED=0 go build -a -o drawio-renderer .
+
+
+FROM node:lts-alpine3.13
```
Comment (Jaakkonen) [inline Dockerfile:7]: Is there version without alpine version pinned? That'd be a bit batter imo

File: README.md [hunk 14]
```diff
@@ -1,29 +1,123 @@
...
+**Optional:** What log level to use. Recognized levels are "info" and "debug".
+
+**Default:** `info`
+
+Examples:
+
+- `info`
+- `debug`
+
+## Usage Example
```
Comment (Jaakkonen) [inline README.md:52]: Imo this long thing should be in separate MD file

**Comment by Jaakkonen** [comment]
Looking quite good to me. File the improvement ideas as new issues or fix them in this PR and I'm approving this.

**Comment by Jaakkonen** [comment]
Maybe document most important use-cases and howtos in README.md or link to file documenting that. I.e. adding new diagram (how to get xml), debugging build locally (copypasta commands, now docker example shows help.
Didn't have time to check that out but I think a user should feel like they know how to use the thing from first 30 seconds reading the readme. It's close to that but maybe could be improved?

**Comment by luxas** [comment]
This PR should be ready to merge now. Docs are added to the readme as well (with pictures)

**Comment by luxas** [comment]
When you approve I'll squash, merge and call this v1

**Review by twelho — APPROVED** [review]
LGTM :+1: just a couple of nits and questions. Feel free to merge after addressing.

`docs/images/7.png` maybe confusingly refers to `special-computing-machine`, you might want to re-render with some `racklet` repo if you see fit, but I'm also fine with this if you see it as just an example. See other comments below.

**Comment by luxas** [comment]
> docs/images/7.png maybe confusingly refers to special-computing-machine, you might want to re-render with some racklet repo if you see fit, but I'm also fine with this if you see it as just an example

Yeah, I see it more as an example. I'll clarify that extra in the README :+1:

File: README.md (hunk 5 above)
Comment (luxas) [reply README.md:167]: yep, now fixed :+1:

File: Makefile (hunk 4 above)
Comment (luxas) [reply Makefile:35]: Strange; the diff didn't show it correctly really by default, I did it now however, so it'll be included in v1.

**Comment by luxas** [comment]
Now I've squashed the commits; so this is ready to merge. Thanks for the review :+1:

**Pull request merged**